primitives/src/merkle_tree/namespaced_merkle_tree.rs

```diff
@@ -181,10 +188,387 @@ where
+    T: NodeValue,
+    H: DigestAlgorithm<E, u64, T> + BindNamespace<E, u64, T, N>,
+    N: Namespace,
+    Arity: Unsigned,
+{
+    fn verify_left_namespace_boundary(
+        &self,
+        root: &NamespacedHash<T, N>,
+        namespace: N,
+    ) -> Result<VerificationResult, PrimitivesError> {
+        if let Some(boundary_proof) = self.left_boundary_proof.as_ref() {
+            // If there is a leaf to the left of the namespace range, check that it is less
+            // than the target namespace
+            if boundary_proof.elem().unwrap().get_namespace() >= namespace
+                || *boundary_proof.index() != self.indices[0] - 1
+            {
+                return Ok(Err(()));
+            }
+        } else {
+            // If there is no left boundary, ensure that target namespace is the tree's
+            // minimum namespace
+            if root.min_namespace != namespace {
+                return Ok(Err(()));
+            }
+        }
+        Ok(Ok(()))
+    }
+
+    fn verify_right_namespace_boundary(
+        &self,
+        root: &NamespacedHash<T, N>,
+        namespace: N,
+    ) -> Result<VerificationResult, PrimitivesError> {
+        if let Some(boundary_proof) = self.right_boundary_proof.as_ref() {
+            // If there is a leaf to the left of the namespace range, check that it is less
+            // than the target namespace
+            if boundary_proof.elem().unwrap().get_namespace() <= namespace
+                || *boundary_proof.index() != self.indices[self.indices.len() - 1] + 1
+            {
+                return Ok(Err(()));
+            }
+        } else {
+            // If there is no left boundary, ensure that target namespace is the tree's
+            // minimum namespace
+            if root.max_namespace != namespace {
+                return Ok(Err(()));
+            }
+        }
+        Ok(Ok(()))
+    }
+
+    /// Verify a namespace proof
+    fn verify(
+        &self,
+        root: &NamespacedHash<T, N>,
+        namespace: N,
+    ) -> Result<VerificationResult, PrimitivesError> {
+        match self.proof_type {
+            NamespaceProofType::Presence => self.verify_presence_proof(root, namespace),
+            NamespaceProofType::Absence => self.verify_absence_proof(root, namespace),
+        }
+    }
+
+    fn verify_absence_proof(
+        &self,
+        root: &NamespacedHash<T, N>,
+        namespace: N,
+    ) -> Result<VerificationResult, PrimitivesError> {
+        if namespace < root.min_namespace || namespace > root.max_namespace {
+            // Easy case where the namespace isn't covered by the range of the tree root
+            return Ok(Ok(()));
+        } else {
+            // Harder case: Find an element whose namespace is greater than our
+            // target and show that the namespace to the left is less than our
+            // target
+            let left_proof = &self.left_boundary_proof.as_ref().unwrap();
```

Comment:
We should return an error if `left_boundary_proof` is `None`, not panic. The proof is untrusted data, the prover should not be able to make the verifier panic.

Reply:
@ggutoski would it be more appropriate to return a `PrimitivesError` here or a falsey `VerificationResult`? The context is that the proof being verified is malformed.